test/appsignal/span_test.exs

```diff
@@ -519,6 +537,177 @@ defmodule AppsignalSpanTest do
     end
   end
 
+  describe ".set_sample_data_if_nil/3" do
+    setup :create_root_span
+
+    setup %{span: span} do
+      [return: Span.set_sample_data_if_nil(span, "key", %{foo: "bar"})]
+    end
+
+    test "returns the span", %{span: span, return: return} do
+      assert return == span
+    end
+
+    @tag :skip_env_test_no_nif
+    test "sets the sample data", %{span: span} do
+      assert %{"sample_data" => %{"key" => "{\"foo\":\"bar\"}"}} = Span.to_map(span)
+    end
+  end
+
+  describe ".set_sample_data_if_nil/3, when sample data has already been set" do
+    setup :create_root_span
+
+    setup %{span: span} do
+      Span.set_sample_data_if_nil(span, "key", %{foo: "bar"})
+      [return: Span.set_sample_data_if_nil(span, "key", %{baz: "quux"})]
+    end
+
+    test "returns the span", %{span: span, return: return} do
+      assert return == span
+    end
+
+    @tag :skip_env_test_no_nif
+    test "does not override the sample data", %{span: span} do
+      assert %{"sample_data" => %{"key" => "{\"foo\":\"bar\"}"}} = Span.to_map(span)
+    end
+  end
+
+  describe ".set_sample_data_if_nil/3, setting params" do
+    setup :create_root_span
+
+    setup %{span: span} do
+      [return: Span.set_sample_data_if_nil(span, "params", %{foo: "bar"})]
+    end
+
+    test "returns the span", %{span: span, return: return} do
+      assert return == span
+    end
+
+    @tag :skip_env_test_no_nif
+    test "sets the sample data", %{span: span} do
+      assert %{"sample_data" => %{"params" => ~s({"foo":"bar"})}} = Span.to_map(span)
+    end
+  end
+
+  describe ".set_sample_data_if_nil/3, setting session_data" do
+    setup :create_root_span
+
+    setup %{span: span} do
+      [return: Span.set_sample_data_if_nil(span, "session_data", %{foo: "bar"})]
+    end
+
+    test "returns the span", %{span: span, return: return} do
+      assert return == span
+    end
+
+    @tag :skip_env_test_no_nif
+    test "sets the sample data", %{span: span} do
+      assert %{"sample_data" => %{"session_data" => ~s({"foo":"bar"})}} = Span.to_map(span)
+    end
+  end
+
+  describe ".set_sample_data_if_nil/3, if send_params is set to false" do
+    setup :create_root_span
+
+    setup %{span: span} do
+      config = Application.get_env(:appsignal, :config)
+      Application.put_env(:appsignal, :config, %{config | send_params: false})
+
+      try do
+        Span.set_sample_data_if_nil(span, "key", %{foo: "bar"})
+      after
+        Application.put_env(:appsignal, :config, config)
+      end
```

Comment:
You can also use the `with_config` helper for this, so you don't have to shuffle the config around and restore it.

Reply:
Ah, yes! This is a shameless copy-paste of the `set_sample_data` tests, which already do it in this way. I did not pay a lot of attention to how they do it.

I'll look at making this change for both.